### PR TITLE
Implement XLAShardedTensor._spec and test

### PR DESCRIPTION
**Pull request by aws-cph**
Implementing and adding tests for XLAShardedTensor._spec in regards to https://github.com/pytorch/xla/issues/9418.

### CODE REVIEW

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 1]
```diff
@@ -169,6 +199,60 @@ def wrap(elem):
...
+
+    # use existing mesh_shape
+    if hasattr(self, 'mesh_shape') and self.mesh_shape:
+      import torch_xla.runtime as xr
+      device_count = xr.global_runtime_device_count()
+      device_list = list(range(device_count))
+      mesh = DeviceMesh("xla",
+                        torch.tensor(device_list).reshape(self.mesh_shape))
+    else:
+      # default to 1D mesh
```
Comment (jeffhataws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: Maybe this should be an error.

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 2]
```diff
@@ -169,6 +199,60 @@ def wrap(elem):
...
+        # find tensor dimension sharded on this mesh dimension
+        tensor_dim = None
+        for t_dim, m_dim in enumerate(self.partition_spec):
+          if m_dim == mesh_dim:
+            tensor_dim = t_dim
+            break
+        placements.append(
+            Shard(tensor_dim) if tensor_dim is not None else Replicate())
+    else:
+      placements = [Replicate()]
```
Comment (jeffhataws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: Same here. Should be an error.

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 3]
```diff
@@ -651,7 +651,9 @@ def mark_sharding(t: Union[torch.Tensor, XLAShardedTensor], mesh: Mesh,
   op_sharding = mesh.get_op_sharding(partition_spec)
   annotate_func = torch_xla._XLAC._xla_mark_sharding
   annotate_func(unwrap_sharded_tensor(t), op_sharding)
-  return wrap_as_sharded_tensor(t)
+  # Pass mesh and partition spec information for DTensor compatibility
+  return wrap_as_sharded_tensor(
```
Comment (jeffhataws) [inline torch_xla/distributed/spmd/xla_sharding.py:658]: Let's do the same for the other APIs above, like annotate_custom_sharding and enable_manual_sharding?

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 4]
```diff
@@ -169,6 +199,60 @@ def wrap(elem):
...
+    if hasattr(self, '_cached_spec'):
+      return self._cached_spec
+
+    from torch.distributed.tensor._dtensor_spec import DTensorSpec, TensorMeta
+    from torch.distributed.device_mesh import DeviceMesh
+    from torch.distributed.tensor.placement_types import Shard, Replicate
+
+    # use existing mesh_shape
+    if hasattr(self, 'mesh_shape') and self.mesh_shape:
+      import torch_xla.runtime as xr
```
Comment (fhaolinaws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: nit: move the import outside if

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 1 above)
Comment (fhaolinaws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: Why do we have default instead of throwing error? Is it for auto wrapping?

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 5]
```diff
@@ -169,6 +199,60 @@ def wrap(elem):
...
+    """
+    # Return cached spec if available
+    if hasattr(self, '_cached_spec'):
+      return self._cached_spec
+
+    from torch.distributed.tensor._dtensor_spec import DTensorSpec, TensorMeta
+    from torch.distributed.device_mesh import DeviceMesh
+    from torch.distributed.tensor.placement_types import Shard, Replicate
+
+    # use existing mesh_shape
```
Comment (fhaolinaws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py:201]: It's better to extract the conversion into a function and put it around here https://github.com/pytorch/xla/blob/e7dcc7bd000ac7b1c0dd02faa32df0a348cc35db/torch_xla/distributed/spmd/api.py#L49

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 6]
```diff
@@ -169,6 +199,60 @@ def wrap(elem):
...
+    if hasattr(self, 'mesh_shape') and self.mesh_shape:
+      import torch_xla.runtime as xr
+      device_count = xr.global_runtime_device_count()
+      device_list = list(range(device_count))
+      mesh = DeviceMesh("xla",
```
Comment (fhaolinaws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py:205]: We need to take care of mesh dim names, too

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 7]
```diff
@@ -169,6 +199,60 @@ def wrap(elem):
...
+                        torch.tensor(device_list).reshape(self.mesh_shape))
+    else:
+      # default to 1D mesh
+      import torch_xla.runtime as xr
+      device_count = xr.global_runtime_device_count()
+      mesh = DeviceMesh("xla", list(range(device_count)))
+
+    # use existing partition_spec
```
Comment (fhaolinaws) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py:214]: Same as above

File: test/spmd/test_xla_dtensor_spec_conversion.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  def test_mesh_conversion(self):
+    device_count = xr.global_runtime_device_count()
+    original_mesh = DeviceMesh("xla", list(range(device_count)))
+    tensor = torch.randn(50, 50)
+    xla_tensor = distribute_tensor(tensor, original_mesh, [Shard(0)])
+
+    converted_spec = xla_tensor._spec
+
+    assert converted_spec.mesh.device_type == "xla"
+    assert converted_spec.mesh.size() == device_count
```
Comment (fhaolinaws) [inline test/spmd/test_xla_dtensor_spec_conversion.py:64]: May need to assert on mesh size each dim, too

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 9]
```diff
@@ -169,6 +203,53 @@ def wrap(elem):
...
     return rs
 
+  @property
+  def _spec(self):
+    """
+    Convert XLA sharding information to DTensorSpec for DTensor interface compatibility.
+    """
+    # Return cached spec if available
+    if hasattr(self, '_cached_spec'):
+      return self._cached_spec
```
Comment (bfolie) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: What if a call to `wrap_as_sharded_tensor` changes `self.mesh_shape` and/or `self.partition_spec`? Will you still get this cached value even though it's out of date?

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 10]
```diff
@@ -169,6 +203,53 @@ def wrap(elem):
...
+
+    # use existing mesh_shape
+    if hasattr(self, 'mesh_shape') and self.mesh_shape:
```
Comment (bfolie) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: Can `self.mesh_shape` and `self.partition_spec` be set to None at initialization? That removes the need to check `hasattr`. I don't think there's a semantic difference between the attribute not existing and its value being None, is there?

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 11]
```diff
@@ -159,7 +173,27 @@ def unwrap(elem):
...
+          nonlocal mesh_shape, partition_spec
+          if isinstance(x, XLAShardedTensor):
+            if hasattr(x, 'mesh_shape') and x.mesh_shape:
+              mesh_shape = x.mesh_shape
+            if hasattr(x, 'partition_spec') and x.partition_spec:
+              partition_spec = x.partition_spec
+
+        tree_map(find_sharded_info, args)
+        if kwargs:
+          tree_map(find_sharded_info, kwargs)
```
Comment (bfolie) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: I don't have enough experience with this codebase to understand the context. What are *args and **kwargs  in practice and why do we expect they would have sharding information that is relevant for `elem`?

File: test/spmd/test_xla_dtensor_spec_conversion.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    spec2 = xla_tensor._spec
+    second_access_time = time.time() - start_time
+
+    assert spec1 is spec2
+    print(
+        f"First access: {first_access_time:.6f}s, Second access: {second_access_time:.6f}s"
+    )
+    assert second_access_time * 10 < first_access_time, \
+        f"Cached access should be much faster: {first_access_time:.6f}s vs {second_access_time:.6f}s"
+
```
Comment (bfolie) [inline test/spmd/test_xla_dtensor_spec_conversion.py]: These sorts of tests that rely on the wall clock often lead to annoying flakes in my experience. I think it's sufficient to just test that `self._cached_spec` has a permanent value after the first call. If you really want to assert that a certain code path is called you could do something with mocks, but that seems like overkill for this, which is simple to confirm by looking at the code ("if the attribute exists, return it" is the first line of the property getter)

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 11 above)
Comment (bfolie) [reply torch_xla/distributed/spmd/xla_sharded_tensor.py]: Also, is this "`elem` is not an `XLAShardedTensor` but there exists sharding information we want to acquire" path tested?

File: torch_xla/distributed/spmd/xla_sharding.py [hunk 13]
```diff
@@ -755,11 +760,19 @@ def clear_sharding(t: Union[torch.Tensor, XLAShardedTensor]) -> torch.Tensor:
...
-    t: Union[torch.Tensor, XLAShardedTensor]) -> XLAShardedTensor:
+def wrap_as_sharded_tensor(t: Union[torch.Tensor, XLAShardedTensor],
+                           mesh_shape=None,
+                           partition_spec=None) -> XLAShardedTensor:
+  # pass along mesh and partition spec information
   if not isinstance(t, XLAShardedTensor):
-    return XLAShardedTensor(t)
-  return t
+    return XLAShardedTensor(
+        t, mesh_shape=mesh_shape, partition_spec=partition_spec)
```
Comment (bfolie) [inline torch_xla/distributed/spmd/xla_sharding.py:770]: The fact that calling `wrap_as_sharded_tensor` in an `XLAShardedTensor` now does something other than trivially  returning `t` is logically new. Is this resharding logic tested?

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 14]
```diff
@@ -169,6 +203,53 @@ def wrap(elem):
                     func(*tree_map(unwrap, args), **tree_map(unwrap, kwargs)))
     return rs
 
+  @property
+  def _spec(self):
+    """
+    Convert XLA sharding information to DTensorSpec for DTensor interface compatibility.
+    """
+    # Return cached spec if available
+    if hasattr(self, '_cached_spec'):
```
Comment (qihqi) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: can we assign self._cache_spec = None in the constructor; then check self._cached_spec is None here.

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 15]
```diff
@@ -91,10 +95,15 @@ class XLAShardedTensor(torch.Tensor):
   # >> assert len(input.shape) == len(partition_spec)
   partition_spec: Tuple[int, None]
 
-  __slots__ = ['global_tensor']
+  __slots__ = ['global_tensor', 'mesh_shape', 'partition_spec', '_cached_spec']
```
Comment (qihqi) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py:99]: can we get rid of __slots__? is it just for performance?

File: torch_xla/distributed/spmd/xla_sharded_tensor.py [hunk 16]
```diff
@@ -159,7 +173,27 @@ def unwrap(elem):
...
+
+        def find_sharded_info(x):
+          nonlocal mesh_shape, partition_spec
+          if isinstance(x, XLAShardedTensor):
+            if hasattr(x, 'mesh_shape') and x.mesh_shape:
+              mesh_shape = x.mesh_shape
+            if hasattr(x, 'partition_spec') and x.partition_spec:
+              partition_spec = x.partition_spec
+
+        tree_map(find_sharded_info, args)
```
Comment (qihqi) [inline torch_xla/distributed/spmd/xla_sharded_tensor.py]: you can do tree_map_only(type, callable, args) then you can skip the isinstance check inside of teh callable

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 15 above)
Comment (aws-cph) [reply torch_xla/distributed/spmd/xla_sharded_tensor.py:99]: My understanding is that it exists for memory optimization. I've kept the `__slots__` attribute as it was in the original code. Please let me know if you'd prefer to have it removed. thank you so much for your feedback!

File: torch_xla/distributed/spmd/xla_sharding.py (hunk 13 above)
Comment (aws-cph) [reply torch_xla/distributed/spmd/xla_sharding.py:770]: Added a test for this case!

File: test/spmd/test_xla_sharding.py [hunk 17]
```diff
@@ -1162,9 +1162,7 @@ def test_mark_shard_scalar(self):
       self.assertIsInstance(shard.indices, type(Ellipsis))
       self.assertEqual(shard.replica_id, i)
 
-    # It looks like mesh_shape attribute is never implemented.
-    with self.assertRaises(AttributeError):
-      xt.mesh_shape
+    # Removed mesh shape assert now that it has been implemented
```
Comment (bfolie) [inline test/spmd/test_xla_sharding.py]: You can remove this comment. It makes sense in the context of this PR but won't make sense to anyone reading the code later

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 9 above)
Comment (aws-cph) [reply torch_xla/distributed/spmd/xla_sharded_tensor.py]: I added logic to handle cache invalidation in wrap_as_sharded_tensor and some other areas in XLAShardedTensor, and added a test to verify spec cache invalidation. Thank you so much for your feedback!

File: torch_xla/distributed/spmd/xla_sharded_tensor.py (hunk 11 above)
Comment (aws-cph) [reply torch_xla/distributed/spmd/xla_sharded_tensor.py]: Thanks for your comments! After reviewing them, I realized that supporting auto-wrapped tensors with sharding information would be more complex than initially anticipated. I ended up updating the PR to revert this change.

File: test/spmd/test_xla_dtensor_spec_conversion.py [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+    """
+    device_count = xr.global_runtime_device_count()
+    mesh = DeviceMesh("xla", torch.arange(device_count))
+    tensor = torch.randn(4, 4)
+    sharded_tensor = distribute_tensor(tensor, mesh, [Shard(0)])
+
+    auto_wrapped = sharded_tensor + sharded_tensor
+
+    with self.assertRaises(ValueError):
+      _ = auto_wrapped._spec
```
Comment (bfolie) [inline test/spmd/test_xla_dtensor_spec_conversion.py:230]: This sort of case, where the behavior is supposed to work but hasn't been implemented yet, is better expressed as an "expected failure". See https://docs.python.org/3/library/unittest.html#skipping-tests-and-expected-failures

**Review by bfolie — APPROVED** [review]
Looks good overall. I have one suggestion about test structure but it's non-blocking

**Review by jeffhataws — APPROVED** [review]
